src/internal_modules/roc_sndio/wav_source.cpp

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (c) 2023 Roc Streaming authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_sndio/wav_source.h"
+#include "roc_core/log.h"
+#include "roc_core/panic.h"
+#include "roc_sndio/backend_map.h"
+
+namespace roc {
+namespace sndio {
+
+WavSource::WavSource(core::IArena& arena, const Config& config)
+    : input_name_(arena)
+    , buffer_(arena)
+    , buffer_size_(0)
+    , eof_(false)
+    , paused_(false)
+    , valid_(false) {
+    BackendMap::instance();
+
+    if (config.sample_spec.num_channels() == 0) {
+        roc_log(LogError, "wav source: # of channels is zero");
+        return;
+    }
+
+    if (config.latency != 0) {
+        roc_log(LogError, "wav source: setting io latency not supported by wav backend");
+        return;
+    }
+
+    frame_length_ = config.frame_length;
+
+    if (frame_length_ == 0) {
+        roc_log(LogError, "wav source: frame length is zero");
+        return;
+    }
+
+    valid_ = true;
+}
+
+WavSource::~WavSource() {
+    close_();
+}
+
+bool WavSource::is_valid() const {
+    return valid_;
+}
+
+bool WavSource::open(const char* path) {
+    roc_panic_if(!valid_);
+
+    roc_log(LogInfo, "wav source: opening: path=%s", path);
+
+    if (file_opened_) {
+        roc_panic("wav source: can't call open() more than once");
+    }
+
+    if (!setup_names_(path)) {
+        return false;
+    }
+
+    if (!open_()) {
+        return false;
+    }
+
+    if (!setup_buffer_()) {
+        return false;
+    }
+
+    file_opened_ = true;
+    return true;
+}
+
+DeviceType WavSource::type() const {
+    return DeviceType_Source;
+}
+
+DeviceState WavSource::state() const {
+    roc_panic_if(!valid_);
+
+    if (paused_) {
+        return DeviceState_Paused;
+    } else {
+        return DeviceState_Active;
+    }
+}
+
+void WavSource::pause() {
+    // no-op - but the state is updated
+    paused_ = true;
+}
+
+bool WavSource::resume() {
+    // no-op - but the state is updated
+    paused_ = false;
+    return true;
+}
+
+bool WavSource::restart() {
+    roc_panic_if(!valid_);
+
+    roc_log(LogDebug, "wav source: restarting: input=%s", input_name_.c_str());
+
+    if (!seek_(0)) {
```

Comment:
Let's panic if file_opened_ is false, otherwise seek may crash since drwav is uninitialized.